In [2]:
import pandas as pd

In [4]:
#Leemos el Dataset
df = pd.read_csv('bank-additional-full.csv',sep = ';')
df

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

In [5]:
df.dtypes

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

In [6]:
# Dataframe sin la ultima columna
x=df.iloc[:, :-1].copy(deep=True)
x.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  duration  campaign  pdays  previous     poutcome  \
0   may         mon       261         1    999         0  nonexistent   
1   may         mon       149         1    999         0  nonexistent   
2   may         mon       226         1    999         0  nonexistent   
3   may         mon       151         1    999         0  nonexistent   
4   may         mon       307         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed  
0           1.1          93.994          -36.4      4.857       5191.0  
1           1.1          93.994          -36.4      4.857       5191.0  
2           1.1          93.994          -36.4      4.857       5191.0  
3           1.1          93.994          -36.4      4.857       5191.0  
4           1.1          93.994          -36.4      4.857       5191.0

In [7]:
#Defino mi target. 
#Es una clasificacion binaria
y = df["y"].copy(deep=True)
y

0         no
1         no
2         no
3         no
4         no
        ... 
41183    yes
41184     no
41185     no
41186    yes
41187     no
Name: y, Length: 41188, dtype: object

In [8]:
#Transformamos el dataset en datos numericos
df=pd.get_dummies(data=df, drop_first=True)
df

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       56       261         1    999         0           1.1          93.994   
1       57       149         1    999         0           1.1          93.994   
2       37       226         1    999         0           1.1          93.994   
3       40       151         1    999         0           1.1          93.994   
4       56       307         1    999         0           1.1          93.994   
...    ...       ...       ...    ...       ...           ...             ...   
41183   73       334         1    999         0          -1.1          94.767   
41184   46       383         1    999         0          -1.1          94.767   
41185   56       189         2    999         0          -1.1          94.767   
41186   44       442         1    999         0          -1.1          94.767   
41187   74       239         3    999         1          -1.1          94.767   

       cons.conf.idx  euribor3m  nr.employed  ...  month_nov  month_oct  \
0              -36.4      4.857       5191.0  ...          0          0   
1              -36.4      4.857       5191.0  ...          0          0   
2              -36.4      4.857       5191.0  ...          0          0   
3              -36.4      4.857       5191.0  ...          0          0   
4              -36.4      4.857       5191.0  ...          0          0   
...              ...        ...          ...  ...        ...        ...   
41183          -50.8      1.028       4963.6  ...          1          0   
41184          -50.8      1.028       4963.6  ...          1          0   
41185          -50.8      1.028       4963.6  ...          1          0   
41186          -50.8      1.028       4963.6  ...          1          0   
41187          -50.8      1.028       4963.6  ...          1          0   

       month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0              0                1                0                0   
1              0                1                0                0   
2              0                1                0                0   
3              0                1                0                0   
4              0                1                0                0   
...          ...              ...              ...              ...   
41183          0                0                0                0   
41184          0                0                0                0   
41185          0                0                0                0   
41186          0                0                0                0   
41187          0                0                0                0   

       day_of_week_wed  poutcome_nonexistent  poutcome_success  y_yes  
0                    0                     1                 0      0  
1                    0                     1                 0      0  
2                    0                     1                 0      0  
3                    0                     1                 0      0  
4                    0                     1                 0      0  
...                ...                   ...               ...    ...  
41183                0                     1                 0      1  
41184                0                     1                 0      0  
41185                0                     1                 0      0  
41186                0                     1                 0      1  
41187                0                     0                 0      0  

[41188 rows x 54 columns]

In [9]:
df.drop(columns='y_yes')
df.y_yes

0        0
1        0
2        0
3        0
4        0
        ..
41183    1
41184    0
41185    0
41186    1
41187    0
Name: y_yes, Length: 41188, dtype: uint8

In [10]:
#Dividimos las variables independientes y la dependiente
X = df.loc[:,df.columns != 'y_yes']
y = df.loc[:,'y_yes']

In [11]:
X

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       56       261         1    999         0           1.1          93.994   
1       57       149         1    999         0           1.1          93.994   
2       37       226         1    999         0           1.1          93.994   
3       40       151         1    999         0           1.1          93.994   
4       56       307         1    999         0           1.1          93.994   
...    ...       ...       ...    ...       ...           ...             ...   
41183   73       334         1    999         0          -1.1          94.767   
41184   46       383         1    999         0          -1.1          94.767   
41185   56       189         2    999         0          -1.1          94.767   
41186   44       442         1    999         0          -1.1          94.767   
41187   74       239         3    999         1          -1.1          94.767   

       cons.conf.idx  euribor3m  nr.employed  ...  month_may  month_nov  \
0              -36.4      4.857       5191.0  ...          1          0   
1              -36.4      4.857       5191.0  ...          1          0   
2              -36.4      4.857       5191.0  ...          1          0   
3              -36.4      4.857       5191.0  ...          1          0   
4              -36.4      4.857       5191.0  ...          1          0   
...              ...        ...          ...  ...        ...        ...   
41183          -50.8      1.028       4963.6  ...          0          1   
41184          -50.8      1.028       4963.6  ...          0          1   
41185          -50.8      1.028       4963.6  ...          0          1   
41186          -50.8      1.028       4963.6  ...          0          1   
41187          -50.8      1.028       4963.6  ...          0          1   

       month_oct  month_sep  day_of_week_mon  day_of_week_thu  \
0              0          0                1                0   
1              0          0                1                0   
2              0          0                1                0   
3              0          0                1                0   
4              0          0                1                0   
...          ...        ...              ...              ...   
41183          0          0                0                0   
41184          0          0                0                0   
41185          0          0                0                0   
41186          0          0                0                0   
41187          0          0                0                0   

       day_of_week_tue  day_of_week_wed  poutcome_nonexistent  \
0                    0                0                     1   
1                    0                0                     1   
2                    0                0                     1   
3                    0                0                     1   
4                    0                0                     1   
...                ...              ...                   ...   
41183                0                0                     1   
41184                0                0                     1   
41185                0                0                     1   
41186                0                0                     1   
41187                0                0                     0   

       poutcome_success  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
41183                 0  
41184                 0  
41185                 0  
41186                 0  
41187                 0  

[41188 rows x 53 columns]

In [12]:
y

0        0
1        0
2        0
3        0
4        0
        ..
41183    1
41184    0
41185    0
41186    1
41187    0
Name: y_yes, Length: 41188, dtype: uint8

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
#Divimos nuestro Dataset en Train y Test
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25)
X_train

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
39745   36      1178         1      6         2          -1.8          93.876   
569     43        87         2    999         0           1.1          93.994   
5982    46       147         1    999         0           1.1          93.994   
5101    46       225         4    999         0           1.1          93.994   
11683   34       205         5    999         0           1.4          94.465   
...    ...       ...       ...    ...       ...           ...             ...   
27100   58       110         2    999         1          -0.1          93.200   
21209   34       111         4    999         0           1.4          93.444   
10012   43        93         2    999         0           1.4          94.465   
10123   39        72         1    999         0           1.4          94.465   
40552   27       415         5     12         1          -1.7          94.027   

       cons.conf.idx  euribor3m  nr.employed  ...  month_may  month_nov  \
39745          -40.0      0.699       5008.7  ...          1          0   
569            -36.4      4.857       5191.0  ...          1          0   
5982           -36.4      4.857       5191.0  ...          1          0   
5101           -36.4      4.857       5191.0  ...          1          0   
11683          -41.8      4.959       5228.1  ...          0          0   
...              ...        ...          ...  ...        ...        ...   
27100          -42.0      4.021       5195.8  ...          0          1   
21209          -36.1      4.963       5228.1  ...          0          0   
10012          -41.8      4.959       5228.1  ...          0          0   
10123          -41.8      4.958       5228.1  ...          0          0   
40552          -38.3      0.886       4991.6  ...          0          0   

       month_oct  month_sep  day_of_week_mon  day_of_week_thu  \
39745          0          0                0                1   
569            0          0                0                0   
5982           0          0                0                0   
5101           0          0                0                0   
11683          0          0                0                0   
...          ...        ...              ...              ...   
27100          0          0                0                0   
21209          0          0                1                0   
10012          0          0                0                0   
10123          0          0                0                1   
40552          0          0                0                0   

       day_of_week_tue  day_of_week_wed  poutcome_nonexistent  \
39745                0                0                     0   
569                  1                0                     1   
5982                 1                0                     1   
5101                 0                0                     1   
11683                0                0                     1   
...                ...              ...                   ...   
27100                0                0                     0   
21209                0                0                     1   
10012                0                1                     1   
10123                0                0                     1   
40552                1                0                     0   

       poutcome_success  
39745                 1  
569                   0  
5982                  0  
5101                  0  
11683                 0  
...                 ...  
27100                 0  
21209                 0  
10012                 0  
10123                 0  
40552                 1  

[30891 rows x 53 columns]

In [17]:
X_test

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
965     47       216         2    999         0           1.1          93.994   
34138   32      1805         4    999         0          -1.8          92.893   
355     43       404         4    999         0           1.1          93.994   
28964   38       351         1    999         1          -1.8          93.075   
3375    28       484         8    999         0           1.1          93.994   
...    ...       ...       ...    ...       ...           ...             ...   
1715    45       545         2    999         0           1.1          93.994   
9606    36       112         4    999         0           1.4          94.465   
25558   43       239         1    999         0          -0.1          93.200   
5304    44       147        42    999         0           1.1          93.994   
12082   42       102         1    999         0           1.4          94.465   

       cons.conf.idx  euribor3m  nr.employed  ...  month_may  month_nov  \
965            -36.4      4.856       5191.0  ...          1          0   
34138          -46.2      1.281       5099.1  ...          1          0   
355            -36.4      4.857       5191.0  ...          1          0   
28964          -47.1      1.405       5099.1  ...          0          0   
3375           -36.4      4.860       5191.0  ...          1          0   
...              ...        ...          ...  ...        ...        ...   
1715           -36.4      4.855       5191.0  ...          1          0   
9606           -41.8      4.961       5228.1  ...          0          0   
25558          -42.0      4.120       5195.8  ...          0          1   
5304           -36.4      4.857       5191.0  ...          1          0   
12082          -41.8      4.947       5228.1  ...          0          0   

       month_oct  month_sep  day_of_week_mon  day_of_week_thu  \
965            0          0                0                0   
34138          0          0                0                0   
355            0          0                1                0   
28964          0          0                0                0   
3375           0          0                0                1   
...          ...        ...              ...              ...   
1715           0          0                0                0   
9606           0          0                1                0   
25558          0          0                0                0   
5304           0          0                0                0   
12082          0          0                0                0   

       day_of_week_tue  day_of_week_wed  poutcome_nonexistent  \
965                  0                1                     1   
34138                0                1                     1   
355                  0                0                     1   
28964                0                0                     0   
3375                 0                0                     1   
...                ...              ...                   ...   
1715                 0                0                     1   
9606                 0                0                     1   
25558                0                1                     1   
5304                 0                0                     1   
12082                0                0                     1   

       poutcome_success  
965                   0  
34138                 0  
355                   0  
28964                 0  
3375                  0  
...                 ...  
1715                  0  
9606                  0  
25558                 0  
5304                  0  
12082                 0  

[10297 rows x 53 columns]

In [18]:
y_train

39745    1
569      0
5982     0
5101     0
11683    0
        ..
27100    0
21209    0
10012    0
10123    0
40552    1
Name: y_yes, Length: 30891, dtype: uint8

In [19]:
y_test

965      0
34138    0
355      0
28964    0
3375     0
        ..
1715     0
9606     0
25558    0
5304     0
12082    0
Name: y_yes, Length: 10297, dtype: uint8

## Decision Tree

In [20]:
from sklearn.tree import DecisionTreeClassifier

In [43]:
#Instanciamos el Modelo
dt = DecisionTreeClassifier()

In [44]:
#Entremos el Modelo
dt.fit(X_train,y_train)

DecisionTreeClassifier()

In [45]:
y_pred_dt=dt.predict(X_test)

In [46]:
y_test

965      0
34138    0
355      0
28964    0
3375     0
        ..
1715     0
9606     0
25558    0
5304     0
12082    0
Name: y_yes, Length: 10297, dtype: uint8

In [47]:
from sklearn.metrics import confusion_matrix

In [48]:
#Calculamos la Matriz de confusion
confusion_dt=confusion_matrix(y_test,y_pred_dt)

In [49]:
confusion_dt

array([[8496,  626],
       [ 531,  644]], dtype=int64)

In [50]:
from sklearn.metrics import accuracy_score

In [51]:
#Calculamos el score de nuestro Modelo
acc_dt=accuracy_score(y_test,y_pred_dt)

In [52]:
acc_dt

0.8876371758764688

<div class="alert alert-block alert-success">
<b> Vemos que el modelo Decision Tree tiene una precision de 88.76% </b> 
</div>

In [32]:
from sklearn.neighbors import KNeighborsClassifier # KNN
from sklearn.linear_model import LogisticRegression # Regresion Logistica
from sklearn.ensemble import RandomForestClassifier # Random Forest

## KNN, regresión logística y random forest

In [72]:
#Inicializamos los modelos
knn = KNeighborsClassifier() # neighbors = 5 (DEFAULT)
lgr = LogisticRegression() # (DEFAULT)
rfc = RandomForestClassifier() # trees = 100 (DEFAULT)

In [35]:
#Entrenamos el modelo KNN
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [39]:
#Precedimos usando el modelo KNN entrenado
y_pred_knn = knn.predict(X_test)

In [41]:
#Calculamos la Matriz de confusion
confusion_knn = confusion_matrix(y_test,y_pred_knn)

In [42]:
#Calculamos el score de nuestro Modelo
acc_knn=accuracy_score(y_test,y_pred_knn)
acc_knn

0.9023016412547344

<div class="alert alert-block alert-success">
<b> Vemos que el modelo KNN tiene una precision de 90.2% </b> 
</div>

In [53]:
#Entrenamos el modelo Logistic Regression
lgr.fit(X_train, y_train)

C:\Users\Lucio\miniconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [54]:
#Precedimos usando el modelo LGR entrenado
y_pred_lgr = lgr.predict(X_test)

In [55]:
#Calculamos la Matriz de confusion
confusion_lgr = confusion_matrix(y_test,y_pred_lgr)

In [56]:
#Calculamos el score de nuestro Modelo
acc_lgr = accuracy_score(y_test,y_pred_lgr)
acc_lgr

0.9058949208507332

<div class="alert alert-block alert-success">
<b> Vemos que el modelo Logistic Regression tiene una precision de 90.6% </b> 
</div>

In [73]:
#Entrenamos el modelo Random Forest
rfc.fit(X_train, y_train)

RandomForestClassifier()

In [74]:
#Precedimos usando el modelo Random Forest entrenado
y_pred_rfc = rfc.predict(X_test)

In [75]:
#Calculamos la Matriz de confusion
confusion_rfc = confusion_matrix(y_test,y_pred_rfc)

In [76]:
#Calculamos el score de nuestro Modelo
acc_rfc = accuracy_score(y_test,y_pred_rfc)
acc_rfc

0.9131785957074876

<div class="alert alert-block alert-success">
<b> Vemos que el modelo Random Forest tiene una precision de 91.31% </b> 
</div>

<div class="alert alert-block alert-success">
<b> De los 4 modelos implementados, Random Forest es el que mas precisión obtiene en la predicción con nuestro dataset </b> 
</div>